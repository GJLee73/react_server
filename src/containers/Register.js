/**
 * 회원가입 컨테이너
 * @author G1
 * @logs // 18.2.25
 */
import React from 'react';
import {connect} from 'react-redux';
import {RegisterComponent} from '../components/auth'
import {idRequest,emailRequest,registerRequest} from '../modules/register';//todo

class Register extends React.Component {

	constructor(props){
		super(props);
		this.state={
			idState:'',
			emailState:'',
			idCheck:false,
			emailCheck:false
		}
		this.handleRegister = this.handleRegister.bind(this);
		this.checkId=this.checkId.bind(this)
		this.checkEmail=this.checkEmail.bind(this)
	}
	componentDidMount(){
		console.log(this.props.status);
		console.log(this.props.errorCode)
	}
	/**
	 * 회원가입요청.
	 * @param {object} msg - 회원가입 form이 들어있다.
	 * @desc
	 *  성공시 로그인 페이지로 리다이렉트.
	 */
	handleRegister(msg){
		return this.props.registerRequest(msg).then(
			
			() => {
				console.log(this.props.status)
				if(this.props.status === "SUCCESS"){
					window.Materialize.toast("SUCCESS! Please log in", 2000);
					this.props.history.push('/login');
					return true;
				}
			}
		).catch((err)=>{
			alert('something worng'+err)

			return 0
		
		}
		);
	}
	/**id가 사용 가능한지알아본다
	 * @param {string} id - 아이디.
		 * @var emailState - 사용가능한아이디입니다 / 아이디가 이미 사용중입니다.
	 * 
	 */
	checkId(id){
		console.log(id)
		if(id===''){
			this.setState({idState:'아이디를입력해주세요', idCheck:this.props.idCheck})
			return 0;
		}
		this.props.idRequest(id).then(()=>{
			this.setState({idState:this.props.idState, idCheck:this.props.idCheck})
		}).catch(()=>{
			this.setState({idState:this.props.idState, idCheck:this.props.idCheck})
		})
			
		
		}
		/**
		 * 이메일이 사용 가능한 이메일인지 찾는다.
		 * @param {*} email 
		 * @var emailState - 사용가능한이메일입니다 / 이메일이 이미 사용중입니다.
		 */
	checkEmail(email){
		console.log(email)
		if(email===''){
			this.setState({emailState:'이메일을입력해주세요', emailCheck:this.props.emailCheck})
			return 0;
		}
		this.props.emailRequest(email).then(()=>{
			console.log('Elkdnfk')
			this.setState({emailState:this.props.emailState, emailCheck:this.props.emailCheck})
		}).catch(()=>{
			this.setState({emailState:this.props.emailState, emailCheck:this.props.emailCheck})
		})
	
	}
	render(){
		return (
			<div>
				< RegisterComponent onRegister={this.handleRegister} checkId={this.checkId} checkEmail={this.checkEmail}
									emailStatus={this.state.emailState} idStatus={this.state.idState} idCheck={this.state.idCheck}
									emailCheck={this.state.emailCheck}/>
			</div>
		);
	}
}

const mapStateToProps = (state) => {
	return {
		status: state.register.regist.status,
		errorCode: state.register.regist.error,
		idState: state.register.id.status,
		emailState: state.register.email.status,
		idCheck:state.register.id.check,
		emailCheck:state.register.email.check
	};
};

const mapDispatchtoProps = (dispatch) => {
	return {
		registerRequest: (id, pw) => dispatch(registerRequest(id, pw)),
		idRequest: (id)=>{
			return dispatch(idRequest(id))
		},
		emailRequest: (email)=>{
			return dispatch(emailRequest(email))
		}
	};
};

export default connect(mapStateToProps, mapDispatchtoProps)(Register);