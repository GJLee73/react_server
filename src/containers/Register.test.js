import React from 'react';
import {shallow, mount} from 'enzyme'
import {Register} from './Register'
import {MemoryRouter} from 'react-router-dom'

describe('Register test',()=>{
 var handleRegister = jest.spyOn(Register.prototype,'handleRegister')
 var checkId = jest.spyOn(Register.prototype,'checkId')
 var checkEmail = jest.spyOn(Register.prototype,'checkEmail')
 var checkPhone = jest.spyOn(Register.prototype,'checkPhone')
 var registerRequest = jest.fn(()=>Promise.resolve())
 var idRequest = jest.fn()
 var emailRequest = jest.fn()
 var phoneRequest = jest.fn()
 var history = {push:jest.fn()}
 global.Materialize = {toast:jest.fn((a,b)=>{return a})}
 global.alert =jest.fn()

 var mounting = (props)=>{
     return mount(<MemoryRouter initialEntries ={[{key:'Register'}]}><Register {...props}/></MemoryRouter>)
 }
 var wrapper = shallow(<Register/>)
    describe('function test',()=>{
        describe('handelRegister test ',()=>{
            it('status===Success',async ()=>{
                wrapper.setProps({registerRequest,status:'SUCCESS',history})
                await wrapper.instance().handleRegister()
                expect(registerRequest).toHaveBeenCalledTimes(1)
                expect(global.Materialize.toast).toHaveBeenCalledTimes(1)
                expect(history.push).toHaveBeenCalledTimes(1)
            })
            it('status!=success',async ()=>{
                wrapper.setProps({status:'fail'})
                await wrapper.instance().handleRegister()
                expect(registerRequest).toHaveBeenCalledTimes(2)
                expect(global.Materialize.toast).toHaveBeenCalledTimes(1)
                expect(history.push).toHaveBeenCalledTimes(1)
           
            })
            it('registerRequest throw error',async()=>{
                wrapper.setProps({registerRequest:()=>Promise.reject()})
                await wrapper.instance().handleRegister()
                expect(global.alert).toHaveBeenCalledTimes(1)
            })
        })
        describe('checkId test',()=>{
         it('id===빈칸',()=>{
            wrapper.instance().checkId('')
            expect(wrapper.state().idState).toBe('아이디를입력해주세요')
         })
         it('아이디가 이미 존재할때 ',async()=>{
             wrapper.setProps({idRequest:()=>Promise.reject(),idState:'아이디가 이미 존재합니다.'})
             await wrapper.instance().checkId('g1')
             expect(wrapper.state().idState).toBe('아이디가 이미 존재합니다.')
         })
         it('아이디가 사용할 수 있을때 ',async()=>{
            wrapper.setProps({idRequest:()=>Promise.resolve(),idState:'아이디를 사용할 수 있습니다.'})
            await wrapper.instance().checkId('g1')
            expect(wrapper.state().idState).toBe('아이디를 사용할 수 있습니다.')
        })
        })
        describe('checkEmail test',()=>{
            it('email === 빈칸',()=>{
                wrapper.instance().checkEmail('')
                expect(wrapper.state().emailState).toBe('이메일을입력해주세요')
            })
            it('email이 이미 등록되어있을때',async()=>{
                wrapper.setProps({emailRequest:()=>Promise.reject(),emailState:'이메일이 이미 등록되어있습니다.'})
                await wrapper.instance().checkEmail('g1')
                expect(wrapper.state().emailState).toBe('이메일이 이미 등록되어있습니다.')
            })
            it('email을 사용할 수 있을때',async()=>{
                wrapper.setProps({emailRequest:()=>Promise.resolve(),emailState:'이메일을 사용할 수 있습니다.'})
                await wrapper.instance().checkEmail('g1')
                expect(wrapper.state().emailState).toBe('이메일을 사용할 수 있습니다.')
            })
        })
        describe('checkPhone',()=>{
            it('phone length<12',()=>{
                wrapper.instance().checkPhone('123-456')
                expect(global.alert).toHaveBeenCalledTimes(2)
            })
            it('phone이 이미 사용중일때',()=>{
                wrapper.setProps({phoneRequest:()=>Promise.resolve(), phoneCheck:false})
                wrapper.instance().checkPhone('000-0000-0000')
                expect(wrapper.state().phoneCheck).toBe(false)
            })
            it('phone이 사용 가능할때',async ()=>{
                wrapper.setProps({phoneRequest:()=>Promise.resolve(), phoneCheck:true})
                await wrapper.instance().checkPhone('000-0000-0000')
                expect(wrapper.state().phoneCheck).toBe(true)
            })
        })
    })
    it('snapShot test',()=>{
        var mounted = mounting({})
        expect(mounted).toMatchSnapshot()
    })
})