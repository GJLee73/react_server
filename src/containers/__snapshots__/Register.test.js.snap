// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register test snapShot test 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "Register",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "Register",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "Register",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Register>
      <div>
        <RegisterComponent
          checkEmail={[Function]}
          checkId={[Function]}
          checkPhone={[Function]}
          emailCheck={false}
          emailStatus=""
          idCheck={false}
          idStatus=""
          onRegister={[Function]}
          phoneCheck={false}
        >
          <div
            className="container auth"
          >
            <Link
              className="logo"
              replace={false}
              to="/"
            >
              <a
                className="logo"
                href="/"
                onClick={[Function]}
              >
                5437
              </a>
            </Link>
            <div
              className="card"
            >
              <div
                className="header blue white-text center"
              >
                <div
                  className="card-content"
                >
                  REGISTER
                </div>
              </div>
              <div>
                <div
                  className="card-content"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="input-field col s12 username"
                    >
                      <label>
                        Id
                      </label>
                      <i />
                      <input
                        className="validate"
                        name="userId"
                        onBlur={[Function]}
                        onChange={[Function]}
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      className="input-field col s12"
                    >
                      <label>
                        Password 
                      </label>
                      <i>
                        비밀번호는 8자이상 최소 한개 이상의 특수문자가 있어야합니다.
                      </i>
                      <input
                        className="validate"
                        name="password"
                        onChange={[Function]}
                        onKeyUp={[Function]}
                        type="password"
                        value=""
                      />
                    </div>
                    <div
                      className="input-field col s12"
                    >
                      <label>
                        Passsword check
                      </label>
                      <i />
                      <input
                        className="validate"
                        name="passwordCheck"
                        onChange={[Function]}
                        onKeyUp={[Function]}
                        type="password"
                        value=""
                      />
                    </div>
                    <div
                      className="input-field col s12 "
                    >
                      <label>
                        이름
                      </label>
                      <input
                        className="validate"
                        name="name"
                        onChange={[Function]}
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      className="input-field col s12 "
                    >
                      <label>
                        생년월일
                      </label>
                      <input
                        className="validate"
                        name="birth"
                        onChange={[Function]}
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      className="input-field col s12 "
                    >
                      <select
                        name="gender"
                        onChange={[Function]}
                        value=""
                      >
                        <option
                          disabled={true}
                          value=""
                        >
                          성별
                        </option>
                        <option
                          value="M"
                        >
                          남자
                        </option>
                        <option
                          value="F"
                        >
                          여자
                        </option>
                      </select>
                    </div>
                    <div
                      className="input-field col s12 "
                    >
                      <label>
                        이메일
                      </label>
                      <i />
                      <input
                        className="validate"
                        name="email"
                        onChange={[Function]}
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      className="input-field col s12 "
                    >
                      <label />
                      <button
                        onClick={[Function]}
                      >
                        이메일체크
                      </button>
                    </div>
                    <div
                      className="input-field col s12 "
                    >
                      <label>
                        휴대폰
                      </label>
                      <input
                        className="validate"
                        maxLength="13"
                        name="phone"
                        onChange={[Function]}
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      className="input-field col s12 "
                    >
                      <label />
                      <button
                        onClick={[Function]}
                      >
                        폰번호체크
                      </button>
                    </div>
                    <a
                      className="waves-effect waves-light btn"
                      onClick={[Function]}
                    >
                      CREATE
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </RegisterComponent>
      </div>
    </Register>
  </Router>
</MemoryRouter>
`;
